### DISCUSSION TITLE
[FR] Site group-based templates

### DISCUSSION THREAD

**Discussion by hongc-cc** [craftcms/cms#8564] [Ideas] [answered]
### Description

If there is only translation between localized site, we probably dont need to add new templates
````
templates/
├── account.twig      --> default site A account template, and works for both EN, DE, FR localization
````
But with multiple sites + localization, 
Site A, Site B, Site C can have a large difference in page layout.
it becomes:

````
templates/
├── account.twig      --> default site A account template,  works for all EN, DE, FR localization
└── siteBen /
    └── account.twig  --> siteB EN account template
└── siteBde /
    └── account.twig  --> siteB DE account template
└── siteBfr /
    └── account.twig  --> siteB FR account template
└── siteCen /
    └── account.twig  --> siteC EN account template
└── siteCde /
    └── account.twig  --> siteC DE account template
└── siteCfr /
    └── account.twig  --> siteC FR account template
````

Can we make use of siteGroup concept, and workout something like this:
````
templates/
├── account.twig      --> all sites default account template,  works for all EN, DE, FR localization
└── siteGroupB /
    └── account.twig  --> siteGroupB account default template,  works for all EN, DE, FR localization
└── siteGroupC /
    └── account.twig  --> siteGroupC account default template,  works for all EN, DE, FR localization
````

**Comment by brandonkelly** [comment]
Seems reasonable!

**Comment by BenParizek** [comment]
Just to note it here. This sounds similar to the need we had for Site Groups with a plugin supporting multi-lingual sitemaps. Nice to see another use case appear for those Site Groups.

**Comment by naboo** [comment]
+1

Also give Site Groups some sort of handle. That is also set on the `{{ currentSite }}` global for use in twig.

**Comment by el-schneider** [comment]
+1 

I would love to see this functionality implemented.

**Comment by WHITE-developer** [comment]
+1 
Exactly in that situation right now, multiple sites with multiple languages which need slightly changes across the sites aka site groups.

**Comment by jan-dh** [comment]
Looking forward to this, wondering if there's a clean work-around in the current setup.

**Comment by jan-dh** [comment]
There's a [PR](https://github.com/ThomasDeMarez/cms/commit/aa5cee4a386576d4b8a1d0b82a327040bca17481) for this

**Comment by zehfred** [comment]
Has anything changed since 2018? I'd like to have templates per site group as well. Can't find any info on the docs on how to accomplish this, so I suppose it hasn't been developed?

**Reply by brandonkelly** [reply]
Not yet. We would need to add handles to site groups for it to be possible.

**Reply by jan-dh** [reply]
There was a [PR](https://github.com/craftcms/cms/pull/6572) for this I think

**Reply by zehfred** [reply]
Thanks :)

**Comment by jannisborgers** [comment]
Just here to add my +1

We have a multi-site installation that will have two sites added annually from now on. 2020 in German and 2020 in English are already in place, now 2021 in both languages will be added. Templates will change across these two new sites, but they will be localized only between each other. If we have to duplicate the templates into de2021 and en2021 template folders, we'd be at 4 duplicates come 2022.

Due to it being legacy code, some tricks (like including/extending by site group) won’t work without me having to touch all the default (2020) code.

I’m looking for workarounds, but having a regular way to overwrite templates on a site **group** basis would solve it perfectly.

**Comment by jan-dh** [comment]
I'm looking at Craft 4 docs to see if anything has changed on this but as far as I can tell it has not?

**Comment by jan-dh** [comment]
So a small update, seems like the Craft 4.0 label was removed [10 days ago](https://github.com/craftcms/cms/pull/6572), had a look at the [5.0 roadmap](https://craftcms.com/roadmap) but it isn't mentioned on it anywhere. Curious to see if this is something that's still on the radar. Was hoping to use this feature in some upcoming project. Some insights on timing and/or wether or not this is ever going to be implemented would be nice.

**Reply by brandonkelly** [reply]
It didn’t make the cut for 4.0 unfortunately. It’s in the backlog, but not planned for a specific release at the moment.

**Reply by jan-dh** [reply]
Thanks for the update

**Comment by Anubarak** [comment]
For anyone wanting this feature: you could as well just create a [symlink ](https://en.wikipedia.org/wiki/Symbolic_link) for your other sites. That's what I usually do in that case
```
templates
├─ siteGroupTemplates
├─ SiteA -> symlink  to siteGroupTemplates
├─ SiteB -> symlink  to siteGroupTemplates
```
siteGroupTemplates contains all the real templates, SiteA and SiteB are just empty folders with a link to siteGroupTemplates

**Reply by brandonkelly** [reply]
Brilliant!

**Reply by jannisborgers** [reply]
This is what we did for the site I mentioned some months ago and it works fine, as long as you can create symlinks on the server. 👍

**Comment by brandonkelly** [comment, marked as answer]
We have decided not to implement this feature.

In addition to @Anubarak’s [symlink-based solution](https://github.com/craftcms/cms/discussions/8564#discussioncomment-2944114), another way you can approach this is with a custom `_siteGroup` [template root](https://craftcms.com/docs/4.x/extend/template-roots.html), defined by a bootstrapped module or private plugin:

```php
use Craft;
use craft\events\RegisterTemplateRootsEvent;
use craft\helpers\StringHelper;
use craft\web\View;
use yii\base\Event;

Event::on(
    View::class,
    View::EVENT_REGISTER_SITE_TEMPLATE_ROOTS,
    function(RegisterTemplateRootsEvent $event) {
        $site = Craft::$app->sites->getCurrentSite();
        $group = StringHelper::toCamelCase($site->getGroup()->name);
        $event->roots['_siteGroup'] = Craft::getAlias("@templates/_siteGroups/$group");
    }
);
```

From there, create a `templates/_siteGroups` directory, and additional directories within there named after your site groups (camelCased).


```treeview
templates/
└── _siteGroups/
    ├── groupA/
    │   ├── foo.twig
    │   └── ...
    └── groupB/
        ├── foo.twig
        └── ...
```

Your templates can then reference site group-specific templates via `_siteGroup/foo.twig` (no need to reference the site group itself).

**Reply by jan-dh** [reply]
I've been trying to get this set up with the approach you proposed @brandonkelly but can't seem to get this working.

in the CMS I have the setup for a section to load the template for that section let's say `_pages/section.twig`.
With the about setup I make a template here: `templates/_siteGroups/groupA/_pages/section.twig`.
But the template still goes looking for _pages/section.twig in the template root

I double checked the sitegroup etc. all names seem correct.

**Reply by jan-dh** [reply]
The above solution (not the symlink one but with the `EVENT_REGISTER_SITE_TEMPLATE_ROOTS` only seems to work to include other templates, not to dynamically make a switch between the way a section is set up in the CM and the template it should land on in the templates folder.

**Reply by jan-dh** [reply]
In order to make the above work I guess `CRAFT_TEMPLATES_PATH` should be dynamic somehow I guess?

**Reply by jan-dh** [reply]
Alternatively this would work as well I guess? Or any downsides in taking this approach? Performance?
```
Event::on(
    View::class,
    View::EVENT_BEFORE_RENDER_TEMPLATE,
    function (TemplateEvent $event) {
        $currentSite = Craft::$app->getSites()->getCurrentSite();
        $siteHandle = $currentSite->handle;

        // Get the site group
        $siteGroup = $currentSite->getGroup();
        
        if ($siteGroup) {
            $siteGroupHandle = $siteGroup->handle;
            
            // Construct the new template path
            $newTemplatePath = "_sites/{$siteGroupHandle}/" . $event->template;

            // Check if the new template exists
            if (Craft::$app->getView()->doesTemplateExist($newTemplatePath)) {
                $event->template = $newTemplatePath;
            }
        }
    }
);

**Reply by brandonkelly** [reply]
@jan-dh If you set up the custom template group as demonstrated, you would prefix your template paths with `_siteGroup/` to access those templates.

So instead of `_pages/section.twig`, you would load `_siteGroup/_pages/section.twig`.

**Reply by jan-dh** [reply]
But wasn't that the whole point of this question, that you wouldn't need to adapt the actual section settings an you could just make a split based on a file system?

**Reply by brandonkelly** [reply]
@jan-dh To be clear, you would prepend `_siteGroup/`, and that would dynamically resolve to the folder named after the current site. So yes, a slight modification is needed, but that’s maybe a little cleaner than `_siteGroups/{site.name|camel}/_pages/section.twig`.

**Reply by jan-dh** [reply]
When reading the original proposal:
```
templates/
├── index.twig      --> all sites default index template,  works for all EN, DE, FR localization
└── siteA/
    └── index.twig  --> site A, works for site A
└── siteGroupB /
    └── index.twig  --> siteGroupB index default template,  works for all EN, DE, FR localization
└── siteGroupC /
    └── index.twig  --> siteGroupC index default template,  works for all EN, DE, FR localization
```
I was hoping something like this would be possible ☝️ 
1. Default to the site specific named folder (within template root)
2. Next, go looking in the siteGroup specific folder (thought the setup you showed with the module did something like this)
3. If no files are to be found there, look in the template root
This would allow for a lot of flexibility & granular control in a multi-site setup. Reality is you sometimes have parts you need for site groups and sometimes you have parts you need specifically for sites. Doing all the if/else logic in the templates makes them very bloated.
